File: alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/biochemistry/conditions/TensionPresent.java
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.biochemistry.conditions;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import it.unibo.alchemist.model.Context;
import it.unibo.alchemist.model.biochemistry.EnvironmentSupportingDeformableCells;
import it.unibo.alchemist.model.Node;
import it.unibo.alchemist.model.Reaction;
import it.unibo.alchemist.model.conditions.AbstractCondition;
import it.unibo.alchemist.model.biochemistry.CircularCellProperty;
import it.unibo.alchemist.model.biochemistry.CircularDeformableCellProperty;

import java.util.stream.Stream;

/**
 *
 */
@SuppressFBWarnings("FE_FLOATING_POINT")
public final class TensionPresent extends AbstractCondition<Double> {

    private static final long serialVersionUID = 1L;
    private final EnvironmentSupportingDeformableCells<?> environment;

    /**
     *
     * @param node the node
     * @param environment the environment
     */
    public TensionPresent(final EnvironmentSupportingDeformableCells<?> environment, final Node<Double> node) {
        super(node);
        this.environment = environment;
    }

    @Override
    public TensionPresent cloneCondition(final Node<Double> node, final Reaction<Double> reaction) {
        if (node.asPropertyOrNull(CircularDeformableCellProperty.class) != null) {
            return new TensionPresent(environment, node);
        }
        throw new IllegalArgumentException("Node must have a "
                + CircularDeformableCellProperty.class.getSimpleName());
    }

    @Override
    public Context getContext() {
        return Context.NEIGHBORHOOD;
    }

    @Override
    public double getPropensityContribution() {
        final Node<Double> thisNode = getNode();
        return environment.getNodesWithinRange(thisNode, environment.getMaxDiameterAmongCircularDeformableCells()).stream()
                //.parallel()
                .flatMap(node ->
                    node.asPropertyOrNull(CircularCellProperty.class) != null ? Stream.of(node) : Stream.empty()
                )
                .mapToDouble(node -> {
                    final double maxRn;
                    final double minRn;
                    final double maxRN = thisNode.asProperty(CircularDeformableCellProperty.class)
                            .getMaximumRadius();
                    final double minRN = thisNode.asProperty(CircularDeformableCellProperty.class).getRadius();
                    if (node.asPropertyOrNull(CircularDeformableCellProperty.class) != null) {
                        maxRn = node.asProperty(CircularDeformableCellProperty.class).getMaximumRadius();
                        minRn = node.asProperty(CircularDeformableCellProperty.class).getRadius();
                    } else {
                        maxRn = node.asProperty(CircularCellProperty.class).getRadius();
                        minRn = maxRn;
                    }
                    final double distance = environment.getDistanceBetweenNodes(node, thisNode);
                    if (maxRn + maxRN - distance < 0) {
                        return 0;
                    } else {
                        if (maxRn == minRn && maxRN == minRN) {
                            return 1;
                        } else {
                            return (maxRn + maxRN - distance) / (maxRn + maxRN - minRn - minRN);
                        }
                    }
                })
                .sum();
    }

    @Override
    public boolean isValid() {
        final Node<Double> thisNode = getNode();
        return environment.getNodesWithinRange(thisNode, environment.getMaxDiameterAmongCircularDeformableCells()).stream()
                .parallel()
                .flatMap(n -> n.asPropertyOrNull(CircularCellProperty.class) != null
                        ? Stream.of(n)
                        : Stream.empty())
                .anyMatch(n -> {
                    final double maxDN =  thisNode.asProperty(CircularDeformableCellProperty.class)
                            .getMaximumRadius();
                    if (n.asPropertyOrNull(CircularDeformableCellProperty.class) != null) {
                        return environment.getDistanceBetweenNodes(n, thisNode)
                                < (maxDN + n.asProperty(CircularDeformableCellProperty.class)
                                .getMaximumRadius());
                    } else {
                        return environment.getDistanceBetweenNodes(n, thisNode) < (maxDN
                                + n.asProperty(CircularCellProperty.class).getRadius());
                    }
                });
    }

}

File: alchemist-incarnation-biochemistry/src/main/kotlin/it/unibo/alchemist/model/biochemistry/conditions/TensionPresent.kt
/*
 * Copyright (C) 2010-2026, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.model.biochemistry.conditions

import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.Node.Companion.asProperty
import it.unibo.alchemist.model.Node.Companion.asPropertyOrNull
import it.unibo.alchemist.model.NodeReaction
import it.unibo.alchemist.model.biochemistry.CircularCellProperty
import it.unibo.alchemist.model.biochemistry.CircularDeformableCellProperty
import it.unibo.alchemist.model.biochemistry.EnvironmentSupportingDeformableCells
import it.unibo.alchemist.model.conditions.AbstractCondition
import it.unibo.alchemist.model.observation.Observable
import it.unibo.alchemist.model.observation.ObservableExtensions.ObservableSetExtensions.combineLatest

/** A condition requiring mechanical tension from at least one nearby circular cell. */
class TensionPresent(private val environment: EnvironmentSupportingDeformableCells<*>, node: Node<Double>) :
    AbstractCondition<Double>(node) {

    private val mechanics: Observable<MechanicalState> = environment
        .observeNodesWithinRange(node, environment.maxDiameterAmongCircularDeformableCells)
        .combineLatest(environment::getPosition) { computeMechanicalState() }

    init {
        requireNotNull(node.asPropertyOrNull<Double, CircularDeformableCellProperty>()) {
            "Node must have a ${CircularDeformableCellProperty::class.simpleName}"
        }
        addObservableDependency(mechanics)
        setValidity(mechanics.map(MechanicalState::valid))
    }

    override fun cloneCondition(newNode: Node<Double>, newReaction: NodeReaction<Double>): TensionPresent {
        requireNotNull(newNode.asPropertyOrNull<Double, CircularDeformableCellProperty>()) {
            "Node must have a ${CircularDeformableCellProperty::class.simpleName}"
        }
        return TensionPresent(environment, newNode)
    }

    /** Current tension factor consumed by [it.unibo.alchemist.model.biochemistry.reactions.BiochemicalNodeReaction]. */
    fun getTension(): Double = mechanics.current.tension

    private fun computeMechanicalState(): MechanicalState {
        val thisNode = getNode()
        val thisCell = thisNode.asProperty<Double, CircularDeformableCellProperty>()
        var valid = false
        var totalTension = 0.0
        environment
            .getNodesWithinRange(thisNode, environment.maxDiameterAmongCircularDeformableCells)
            .mapNotNull { neighbor ->
                neighbor.asPropertyOrNull<Double, CircularCellProperty>()?.let { neighbor to it }
            }.forEach { (neighbor, neighborCell) ->
                val distance = environment.getDistanceBetweenNodes(neighbor, thisNode)
                val neighborMaximumRadius =
                    (neighborCell as? CircularDeformableCellProperty)?.maximumRadius ?: neighborCell.radius
                val maximumRadiusSum = thisCell.maximumRadius + neighborMaximumRadius
                valid = valid || distance < maximumRadiusSum
                val currentRadiusSum = thisCell.radius + neighborCell.radius
                totalTension += when {
                    maximumRadiusSum < distance -> 0.0
                    maximumRadiusSum == currentRadiusSum -> 1.0
                    else -> (maximumRadiusSum - distance) / (maximumRadiusSum - currentRadiusSum)
                }
            }
        return MechanicalState(valid, totalTension)
    }

    private data class MechanicalState(val valid: Boolean, val tension: Double)
}
